### ISSUE TITLE
Embedding landmarks into PNG. Better face hull ( non convex )

### ISSUE THREAD

**Issue by iperov** [deepfakes/faceswap-playground#115]
already did  subj.

![out00002_0](https://user-images.githubusercontent.com/8076202/37877447-8aac2a2c-306c-11e8-8bc4-e804ff7af6d0.png)

but I will not release it for a while, because I am making improvements for original model.

**Comment by iperov** [comment]
![winhex_2018-03-27_00-25-11](https://user-images.githubusercontent.com/8076202/37931119-af966336-3155-11e8-8bc6-17a1885742f1.png)

**Comment by iperov** [comment]
also we will no more use alignments.json
all data will be embedded into aligned png files

**Comment by torzdf** [comment]
Will it be possible to extract the information out and write it back in again. Say a false positive, or a face we don't want mapped is identified in a frame and we want to remove it?

**Comment by Apollo122** [comment]
In that case I think simply deleting file will be enough. Because that info is in that extracted png. It's a pretty good PR, we can now sort and delete extracted faces without worrying about alignment.json. My only concern is performance, considering each extracted image will have it's info written itself. Or during convert info must be extracted from image

**Comment by mul5kuq5pn** [comment]
I switched over to JPG files many weeks ago.
Because JPG uses like 10x less disk space.
With clips with 100k or more frames it gets super big very fast.
Please take this into consideration.

**Comment by Apollo122** [comment]
Yeah that's because this will probably be optional along with alignment.json

**Comment by torzdf** [comment]
> In that case I think simply deleting file will be enough. Because that info is in that extracted png. 

Yeah, I agree this is a good feature. In my use case though (and I accept that this is my use case, not everyone's), I don't keep the extracted faces from my target frames, as they just take up unnecessary space. I currently delete false positives/unwanted faces, run a script to edit alignments.json and delete the faces.

I suppose I can just as easily amend my alignments.json script to write straight back to the source pngs, so no biggy.

**Comment by Apollo122** [comment]
As I said this should be optional, for the people who prefer using alignment.json also for backward compatibility

**Comment by bryanlyon** [comment]
I'm not even sure this needs to be optional.  I doubt adding some text data to the PNG while saving takes much time at all.  If so, there is no reason not to do this in addition to the json file by default.

**Comment by kvrooman** [comment]
I was going to suggest saving everything as a TFRecord file with each element composed of 5 components. Good compression and better yet, everything will be pre-optimized to run in parallel on multiple CPUs/GPUs/servers; then we can start using the Tensorflow dataset input pipeline ( which Keras can also directly use now too)

Components
1. the original image/frame  ( type .jpg )
2. the frontal face from running 1. through the alignment chain ( type .jpg )
3. the coordinates of the dlib bounding box  in 1. ( alternatively, a true segmentation mask for 1.( type str)
4. the coordinates of the facial landmarks in 1.( type str)
5. the filename/path of 1. ( type str)

**Comment by iperov** [comment]
`My only concern is performance, considering each extracted image will have its info written itself`
performance is excellent.

**Comment by iperov** [comment]
```
    def prepare_images(self):
        alignments = {} #Collecting alignments
        for filename in tqdm(self.input_aligned_dir, desc="Collecting alignments"):
            a_png = AlignedPNG.load( str(filename) )
            if a_png is None: continue
            d = a_png.getFaceswapDictData()
            if d is None or d['source_filename'] is None or d['source_rect'] is None or d['landmarks'] is None: continue          
            
            if d['source_filename'] not in alignments.keys():
                alignments[ Path(d['source_filename']).stem ] = []
                
            sr = d['source_rect']            
                
            alignments[ Path(d['source_filename']).stem ].append ( 
                DetectedFace ( r=sr['r'], x=sr['x'], w=sr['w'], y=sr['y'], h=sr['h'], landmarksXY=d['landmarks']) )
            
        for filename in tqdm(self.read_directory()):
            image = cv2.imread(filename)

            if not Path(filename).stem in alignments.keys():
                print ('no faces found for %s, skipping' % (os.path.basename(filename)) )
                continue
            
            faces = alignments[ Path(filename).stem ] 
            yield filename, image, faces

```

**Comment by iperov** [comment]
`My only concern is performance, considering each extracted image will have its info written itself`
performance is excellent. 

even 20k files sorting by yaw in 20 secs in every train launch

**Comment by iperov** [comment]
will create new issue in faceswap repo

**Comment by Apollo122** [comment]
@bryanlyon 
If there's no impact on performance and if it will be an addition to alignment.json then yeah it can be default. Some people use .jpg in these case alignment.json may take over

**Comment by iperov** [comment]
@Apollo122 
I fully removed alignments.json

**Comment by Clorr** [comment]
I think embedding data in image file would solve quite a few problems with the alignments.json so I'm fine with that approach.

**Comment by Kirin-kun** [comment]
PNG doesn't support EXIF. I think it's more along [this](https://www.w3.org/TR/PNG/#11textinfo). It's in the PNG specification.

"PNG provides the tEXt, iTXt, and zTXt chunks for storing text strings associated with the image, such as an image description or copyright notice. Keywords are used to indicate what each text string represents. Any number of such text chunks may appear, and more than one with the same keyword is permitted."

The steganographic option might be another way, which would support both png and jpg.

**Comment by iperov** [comment]
pull https://github.com/deepfakes/faceswap/pull/319
